### PR TITLE
undbx: init at 0.22-unstable-2019-02-11

### PR DESCRIPTION
**Pull request by D3vil0p3r**
## Description of changes
Tagged as v0.22 because when the binary is run, the banner shows 0.22.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/un/undbx/package.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    repo = "undbx";
+    rev = "5e31c757e137a6409115cac0623d61d384019b7a";
+    hash = "sha256-leregcv3dv/D3WvFkYyjQePdKi4BgE0aj5PY6JiSKl8=";
+  };
+
+  nativeBuildInputs = [ autoreconfHook ];
+
+  makeFlags = [
+    "CC=cc"
+  ];
```
Comment (Lurkki14) [inline pkgs/by-name/un/undbx/package.nix:22]:
```suggestion
```
I don't see a reason for this

File: pkgs/by-name/un/undbx/package.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  meta = with lib; {
+    description = "Extract e-mail messages from Outlook Express DBX files";
+    homepage = "https://github.com/ZungBang/undbx";
+    maintainers = with maintainers; [ d3vil0p3r ];
+    platforms = platforms.unix;
+    license = licenses.gpl3Plus;
```
Comment (Lurkki14) [inline pkgs/by-name/un/undbx/package.nix:29]:
```suggestion
    license = licenses.gpl3Plus;
    mainProgram = "undbx";
```

**Review by Lurkki14 — CHANGES_REQUESTED** [review]
There also seems to be a broken `undbx.hta` in `$out/bin`

File: pkgs/by-name/un/undbx/package.nix (hunk 1 above)
Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix:22]: Darwin compatibility

Comment (Lurkki14) [reply pkgs/by-name/un/undbx/package.nix:22]: What happens without it? To me this looks like something `autoreconfHook` should handle though, if it is indeed required

Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix:22]: I will test it what happened on Darwin ofborg check

Comment (Lurkki14) [reply pkgs/by-name/un/undbx/package.nix:22]: > Some builds hardcode gcc/g++ in their build scripts, that can usually be fixed with using something like makeFlags = [ "CC=cc" ]; or by patching the build scripts.

Looking at the build files, I don't see hardcoded gcc

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/un/undbx/package.nix:22]: as [Lurkki14](https://github.com/Lurkki14) said, this is unnecessary. the darwin build is failing for different reasons.
 > dbxread.c:178:9: error: variable 'n' set but not used [-Werror,-Wunused-but-set-variable]
    int n;
        ^
dbxread.c:571:31: error: operator '?:' has lower precedence than '+'; '+' will be evaluated first [-Werror,-Wparentheses]
        for (k1 = i + (i < k2)? 1:0;  k0 < k2;  k1 = (k0 + k2) / 2) {
                  ~~~~~~~~~~~~^

this is due to differences in warings between clang and gcc. can reproduce on linux by:
```console
nix-build -A pkgsLLVM.undbx
```

looking at the code it is unclear if `k1 = i + (i < k2)? 1:0` results in a bug or not.

Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix:22]: Ok. I will use only linux platform then.

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/un/undbx/package.nix:22]: the error is not platform related, it is the compiler used and the build using `-Werror` which makes warnings into errors. one fix is to remove `-Werror` flag. usually, i like to report the issue upstream as well.

Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix:22]: You mean if we remove `-Werror` it will compile on Darwin correctly?

Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/un/undbx/package.nix:22]: you can build using clang on linux via `nix-build -A pkgsLLVM.undbx` to test.

Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix]: As agreed with the upstream developer, I applied the patch for it. What I need to understand is to check if that undbx.hta generated on bin directory is needed or not. It seems to be a Visual Basic file, maybe intended for Windows environment.

Comment (D3vil0p3r) [reply pkgs/by-name/un/undbx/package.nix]: The upstream developer confirmed me it is a file intended only for Windows. He suggested me just to comment the line
```
bin_SCRIPTS = undbx.hta
```
in Makefile.am.

File: pkgs/by-name/un/undbx/package.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  nativeBuildInputs = [ autoreconfHook ];
+
+  patches = [
+    # change Makefile.am to allow the GCC hardening.
+    (fetchpatch {
+      url = "https://sources.debian.org/data/main/u/undbx/0.21-4/debian/patches/010_fix-makefile.patch";
+      hash = "sha256-tVrluH96R/LGhCnExaGUALm0BsPQtEuXduItnTHGL38=";
+    })
+  ];
+
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/un/undbx/package.nix:28]:
```suggestion
  postPatch = ''
    substituteInPlace Makefile.am \
      --replace-fail "-Werror" "" \
      --replace-fail "bin_SCRIPTS" "#bin_SCRIPTS"
  '';
```

i am not sure if sources.debian.org urls are stable
https://nixos.org/manual/nixpkgs/unstable/#fetchdebianpatch
suggests they are not.

**Review by ghost — APPROVED** [review]

**Pull request merged**